core/arch/arm/plat-synquacer/rng_pta.c

```diff
@@ -0,0 +1,211 @@
+	0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
+	0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c, 0x8f659eff, 0xf862ae69,
+	0x616bffd3, 0x166ccf45, 0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2,
+	0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db, 0xaed16a4a, 0xd9d65adc,
+	0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
+	0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605, 0xcdd70693,
+	0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
+	0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
+};
+
+static uint32_t crc32(uint8_t *msg, uint32_t sz)
+{
+	uint32_t i, crc = 0xffffffff;
+
+	for (i = 0; i < sz; i++)
+		crc = (crc >> 8) ^ table[(crc ^ msg[i]) & 0xff];
+
+	return ~crc;
+}
+
+static void pool_add_entropy(uint8_t byte)
+{
+	entropy_pool[entropy_size] = byte;
+	entropy_size++;
+}
+
+static void pool_get_entropy(uint8_t *buf, uint32_t size)
+{
+	uint32_t off;
+
+	if (size > entropy_size)
+		return;
+
+	off = entropy_size - size;
+
+	memcpy(buf, &entropy_pool[off], size);
+	entropy_size -= size;
+}
+
+static uint32_t pool_get_entropy_sz(void)
+{
+	uint32_t sz;
+
+	sz = entropy_size;
+
+	return sz;
+}
+
+void rng_collect_entropy(void)
+{
+	uint64_t ts0 = 0, ts1 = 0, ts2 = 0, ts3 = 0, ts4 = 0, ts5 = 0, ts6 = 0;
+	uint8_t e = 0;
+
+	ts0 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE0 +
+					      TEMP_DATA_REG_OFFSET));
+	ts1 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE1 +
+					      TEMP_DATA_REG_OFFSET));
+	ts2 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE2 +
+					      TEMP_DATA_REG_OFFSET));
+	ts3 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE3 +
+					      TEMP_DATA_REG_OFFSET));
+	ts4 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE4 +
+					      TEMP_DATA_REG_OFFSET));
+	ts5 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE5 +
+					      TEMP_DATA_REG_OFFSET));
+	ts6 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE6 +
+					      TEMP_DATA_REG_OFFSET));
+
+	e = ((ts0 & 0x1) << 6) | ((ts1 & 0x1) << 5) | ((ts2 & 0x1) << 4) |
+	    ((ts3 & 0x1) << 3) | ((ts4 & 0x1) << 2) | ((ts5 & 0x1) << 1) |
+	    ((ts6 & 0x1) << 0);
+
+	e |= ((ts0 ^ ts1 ^ ts2 ^ ts3 ^ ts4 ^ ts5 ^ ts6) << 6) & 0x80;
+
+	pool_add_entropy(e);
+
+	if (pool_get_entropy_sz() >= ENTROPY_POOL_SIZE)
+		generic_timer_stop();
+}
+
+static TEE_Result rng_get_entropy(uint32_t types,
+				  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint8_t *e = NULL;
+	uint32_t i, digest, pool_size = 0, rq_size = 0;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	rq_size = params[0].memref.size;
+
+	if ((rq_size == 0) || (rq_size > ENTROPY_POOL_SIZE))
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	e = (uint8_t *)params[0].memref.buffer;
+	if (e == NULL)
```

Comment:
We usually don't compare agains `== NULL`, i.e., I think you will see the negation mostly in the OP-TEE code base.

Reply:
ok will use negation then.